src/object/obj_layout.c

```diff
@@ -58,18 +58,32 @@ obj_layout_diff(struct pl_map *map, daos_unit_oid_t oid, uint32_t new_ver, uint3
 	if (rc)
 		D_GOTO(out, rc);
 
-	if (new_layout->ol_shards[shard].po_target != old_layout->ol_shards[shard].po_target) {
-		*tgt = new_layout->ol_shards[shard].po_target;
-		*shard_p = shard;
-		D_GOTO(out, rc = 1);
-	}
-
 	/* If the new layout changes dkey placement, i.e. dkey->grp, dkey->ec_start changes,
 	 * then all shards needs to be changed.
 	 */
 	if (new_ver == 1 && daos_obj_id_is_ec(oid.id_pub)) {
-		*tgt = new_layout->ol_shards[shard].po_target;
-		*shard_p = shard;
+		struct daos_oclass_attr	*oc;
+		unsigned int	grp_size;
+		unsigned int	grp_start;
+		int		i;
+
+		oc = daos_oclass_attr_find(oid.id_pub, NULL);
+		D_ASSERT(oc != NULL);
+		grp_size = daos_oclass_grp_size(oc);
+
+		D_ASSERT(grp_size < array_size);
+		grp_start = (shard / grp_size) * grp_size;
+		for (i = 0; i < grp_size; i++) {
+			tgts[i] = new_layout->ol_shards[grp_start + i].po_target;
+			shards[i] = grp_start + i;
+			D_DEBUG(DB_TRACE, "i %d tgts[i] %u shards %u grp_size %u\n", i, tgts[i], shards[i], grp_size);
+		}
+		D_GOTO(out, rc = grp_size);
```

Comment:
Sorry I am not quite understand, for other shard (shard[i] != oid.id_shard), other servers/vos_xs will can its own container/object and will find those shards right? then on those other servers/vos_xs if find "new_layout->ol_shards[shard].po_target != old_layout->ol_shards[shard].po_target" will send/rebuild those shards, why here must check for all shards within the group?

Reply:
because dkey->shard mapping has been changed. i.e. dkey might be locate in different shard within the group.  So we can not only check the existing shard, it has to check all of shards within the group.